### PR TITLE
Optimize snakecase processing

### PR DESCRIPTION
**Pull request by deankarn**
**Details**
I was messing around with [Rust](https://www.rust-lang.org/) which has the ability to borrow string data and makes you think a little bit differently and this is a new algorithm I came up with as a result; it avoids allocating and manipulating the string unless necessary.

Probably would want to test in stage first with a larger set of data, but it's here when ready 😄 

**Comparison**
```
benchcmp old.txt new.txt
benchmark                                     old ns/op     new ns/op     delta
BenchmarkSnakecaseBorrowedLong-8              84.9          19.0          -77.62%
BenchmarkSnakecaseBorrowedSimple-8            40.0          8.95          -77.62%
BenchmarkSnakecaseOwnedUnicode-8              41.4          36.7          -11.35%
BenchmarkSnakecaseOwnedLong-8                 71.8          57.2          -20.33%
BenchmarkSnakecaseOwnedLongSpecialChars-8     102           77.1          -24.41%
BenchmarkSnakecaseOwnedSimple-8               39.2          34.1          -13.01%
BenchmarkSnakecaseOwnedUnicode2-8             102           76.5          -25.00%

benchmark                                     old MB/s     new MB/s     speedup
BenchmarkSnakecaseBorrowedLong-8              388.67       1735.43      4.47x
BenchmarkSnakecaseBorrowedSimple-8            274.70       1229.60      4.48x
BenchmarkSnakecaseOwnedUnicode-8              217.49       245.46       1.13x
BenchmarkSnakecaseOwnedLong-8                 404.05       507.17       1.26x
BenchmarkSnakecaseOwnedLongSpecialChars-8     292.22       388.86       1.33x
BenchmarkSnakecaseOwnedSimple-8               280.62       322.92       1.15x
BenchmarkSnakecaseOwnedUnicode2-8             252.78       339.81       1.34x

benchmark                                     old allocs     new allocs     delta
BenchmarkSnakecaseBorrowedLong-8              1              0              -100.00%
BenchmarkSnakecaseBorrowedSimple-8            1              0              -100.00%
BenchmarkSnakecaseOwnedUnicode-8              1              1              +0.00%
BenchmarkSnakecaseOwnedLong-8                 1              1              +0.00%
BenchmarkSnakecaseOwnedLongSpecialChars-8     1              1              +0.00%
BenchmarkSnakecaseOwnedSimple-8               1              1              +0.00%
BenchmarkSnakecaseOwnedUnicode2-8             1              1              +0.00%

benchmark                                     old bytes     new bytes     delta
BenchmarkSnakecaseBorrowedLong-8              48            0             -100.00%
BenchmarkSnakecaseBorrowedSimple-8            16            0             -100.00%
BenchmarkSnakecaseOwnedUnicode-8              3             3             +0.00%
BenchmarkSnakecaseOwnedLong-8                 48            48            +0.00%
BenchmarkSnakecaseOwnedLongSpecialChars-8     32            32            +0.00%
BenchmarkSnakecaseOwnedSimple-8               16            16            +0.00%
BenchmarkSnakecaseOwnedUnicode2-8             3             3             +0.00%
```

**Old**
```
BenchmarkSnakecaseBorrowedLong-8            	20000000	        84.9 ns/op	 388.67 MB/s	      48 B/op	       1 allocs/op
BenchmarkSnakecaseBorrowedSimple-8          	30000000	        40.0 ns/op	 274.70 MB/s	      16 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedUnicode-8            	30000000	        41.4 ns/op	 217.49 MB/s	       3 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedLong-8               	20000000	        71.8 ns/op	 404.05 MB/s	      48 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedLongSpecialChars-8   	20000000	       102 ns/op	 292.22 MB/s	      32 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedSimple-8             	50000000	        39.2 ns/op	 280.62 MB/s	      16 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedUnicode2-8           	20000000	       102 ns/op	 252.78 MB/s	       3 B/op	       1 allocs/op
```

**New**
```
BenchmarkSnakecaseBorrowedLong-8            	100000000	        19.0 ns/op	1735.43 MB/s	       0 B/op	       0 allocs/op
BenchmarkSnakecaseBorrowedSimple-8          	200000000	         8.95 ns/op	1229.60 MB/s	       0 B/op	       0 allocs/op
BenchmarkSnakecaseOwnedUnicode-8            	50000000	        36.7 ns/op	 245.46 MB/s	       3 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedLong-8               	30000000	        57.2 ns/op	 507.17 MB/s	      48 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedLongSpecialChars-8   	20000000	        77.1 ns/op	 388.86 MB/s	      32 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedSimple-8             	50000000	        34.1 ns/op	 322.92 MB/s	      16 B/op	       1 allocs/op
BenchmarkSnakecaseOwnedUnicode2-8           	20000000	        76.5 ns/op	 339.81 MB/s	       3 B/op	       1 allocs/op
```

### CODE REVIEW

File: benchmarks_test.go
Comment (achille-roussel) [inline benchmarks_test.go]: This `ResetTimer` call seems unnecessary, the time elapsed since the benchmark function was called is insignificant compared to the time in the loop.

File: benchmarks_test.go
Comment (achille-roussel) [inline benchmarks_test.go]: This can be enabled on the command line with `-benchmem` ;)

File: snake.go
Comment (achille-roussel) [inline snake.go]: `s/borrow/share` ? "borrowing" is not really a common term in Go land.

File: snake.go
Comment (achille-roussel) [inline snake.go]: How about reversing the logic to reduce the depth of indentation? This usually makes the code easier to read:
```go
if idx == len(s) {
    return s
}

...
```

**Review by achille-roussel — COMMENTED** [review]
This is a cool idea, it's widely used in the Go standard library as well.

I'm a bit concerned about what looks like a behavior change regarding how digits are handled, looping @dominicbarnes who has code which is highly sensitive to the way this snakecase function behaves.

Also looping @maxence-charriere who's used this function recently in another project.

File: benchmarks_test.go
Comment (deankarn) [reply benchmarks_test.go]: Ya Just out of habit

File: snake.go
Comment (deankarn) [reply snake.go]: Ya can change it to whatever :) just using the rust benchmark names so I could compare ;)

**Comment by deankarn** [comment]
Thanks @achille-roussel for the feedback

I saw that number behaviour change and made a fix.

Oh I totally agree under no circumstances should this be merged without extremely thourough testing :)

will make a few adjustments based on your suggestions:)

**Comment by deankarn** [comment]
@dominicbarnes @achille-roussel I was finally able to achieve feature parity which yields the following:
```
benchmark                                   old ns/op     new ns/op     delta
BenchmarkUnchangedLong-8                    95.2          37.6          -60.50%
BenchmarkUnchangedSimple-8                  46.1          14.5          -68.55%
BenchmarkModifiedUnicode-8                  36.4          41.4          +13.74%
BenchmarkModifiedLong-8                     85.6          81.9          -4.32%
BenchmarkModifiedLongSpecialChars-8         128           127           -0.78%
BenchmarkModifiedSimple-8                   42.3          45.0          +6.38%
BenchmarkModifiedUnicode2-8                 71.7          79.4          +10.74%
BenchmarkLeadingUnderscoresDigitUpper-8     37.8          39.4          +4.23%
BenchmarkDigitUpper-8                       33.4          28.9          -13.47%
BenchmarkDigitUpper2-8                      83.5          84.8          +1.56%

benchmark                                   old MB/s     new MB/s     speedup
BenchmarkUnchangedLong-8                    346.67       878.58       2.53x
BenchmarkUnchangedSimple-8                  238.83       757.91       3.17x
BenchmarkModifiedUnicode-8                  247.09       217.28       0.88x
BenchmarkModifiedLong-8                     338.64       354.06       1.05x
BenchmarkModifiedLongSpecialChars-8         233.89       234.43       1.00x
BenchmarkModifiedSimple-8                   259.76       244.48       0.94x
BenchmarkModifiedUnicode2-8                 362.85       327.45       0.90x
BenchmarkLeadingUnderscoresDigitUpper-8     158.70       152.21       0.96x
BenchmarkDigitUpper-8                       149.82       172.80       1.15x
BenchmarkDigitUpper2-8                      203.71       200.45       0.98x

benchmark                                   old allocs     new allocs     delta
BenchmarkUnchangedLong-8                    1              0              -100.00%
BenchmarkUnchangedSimple-8                  1              0              -100.00%
BenchmarkModifiedUnicode-8                  1              1              +0.00%
BenchmarkModifiedLong-8                     1              1              +0.00%
BenchmarkModifiedLongSpecialChars-8         1              1              +0.00%
BenchmarkModifiedSimple-8                   1              1              +0.00%
BenchmarkModifiedUnicode2-8                 1              1              +0.00%
BenchmarkLeadingUnderscoresDigitUpper-8     1              1              +0.00%
BenchmarkDigitUpper-8                       1              1              +0.00%
BenchmarkDigitUpper2-8                      1              1              +0.00%

benchmark                                   old bytes     new bytes     delta
BenchmarkUnchangedLong-8                    48            0             -100.00%
BenchmarkUnchangedSimple-8                  16            0             -100.00%
BenchmarkModifiedUnicode-8                  3             3             +0.00%
BenchmarkModifiedLong-8                     48            48            +0.00%
BenchmarkModifiedLongSpecialChars-8         32            32            +0.00%
BenchmarkModifiedSimple-8                   16            16            +0.00%
BenchmarkModifiedUnicode2-8                 3             3             +0.00%
BenchmarkLeadingUnderscoresDigitUpper-8     5             5             +0.00%
BenchmarkDigitUpper-8                       5             5             +0.00%
BenchmarkDigitUpper2-8                      32            32            +0.00%
```

### Testing
I tested by randomly generating 1 billion random strings, running them through the old logic and then verifying the new logic generated the exact same output; here are the programs for reference:
Generate
```go
package main

import (
	"fmt"
	"math/rand"
	"time"

	snakecase "github.com/segmentio/go-snakecase"
)

const stringLen = 35

var characters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_@ẞ•¶§ƒ∆˚¬")

func init() {
	rand.Seed(time.Now().UnixNano())
}

func getRandomString(b []rune) []rune {
	num := rand.Intn(stringLen)
	if num == 0 {
		num++
	}
	for i := 0; i < num; i++ {
		b = append(b, characters[rand.Intn(len(characters))])
	}
	return b
}

func main() {
	var b []rune
	for i := 0; i < 1000000000; i++ {
		b = getRandomString(b[:0])
		s := string(b)
		fmt.Printf("%s:%s\n", s, snakecase.Snakecase(s))
	}
}
```
Verify
```go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"

	snakecase "github.com/segmentio/go-snakecase"
)

func main() {
	f, err := os.Open("test.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	r := bufio.NewReader(f)
	s := bufio.NewScanner(r)
	var sc string
	for s.Scan() {
		st := strings.Split(s.Text(), ":")
		sc = snakecase.Snakecase(st[0])
		if sc != st[1] {
			fmt.Printf("%s:%s:%s\n", st[0], st[1], sc)
		}
	}
}
```

File: snake.go
Comment (achille-roussel) [inline snake.go:39]: Is 64 still a good choice of size? I used a static size initially because I wanted it to be stack-allocated, but since we know we're gonna have to modify the string and the result will be larger than the original, it's very likely that this buffer will end up being heap allocated.

It may be worth doing `make([]byte, 0, 2 * len(s))` (or some other factor), then instead of `return string(b)` we could do `return *(*string)(unsafe.Pointer(&b))`.
It does add a dependency to `unsafe` which we didn't have before, so possibly increasing the maintenance cost further.

**Comment by achille-roussel** [comment]
It looks like some benchmarks are running slower with this change, would it be worth testing if this is actually yielding positive returns on a real data set?

**Comment by deankarn** [comment]
@achille-roussel on the 1 billion string test which consists of many unicode and changed strings and it still yielded a small increase is performance and that test is almost the worst cases real-world example.

99% of the time from my side it will hit the new optimized paths.

File: snake.go
Comment (deankarn) [reply snake.go:39]: @achille-roussel I did try a few variants including `make([]byte, 0, 2 * len(s))` which yielded much poorer performance on my machine. I think it's due the 64 aligning better to the x86_64 arch even though it's larger.

I do like the suggestion of `return *(*string)(unsafe.Pointer(&b))` though, going to try that and see if there is an increase worth the unsafe dependency.

Comment (deankarn) [reply snake.go:39]: here is the result with `make([]byte, 0, 2 * len(s))`:
```
benchmark                                   old ns/op     new ns/op     delta
BenchmarkUnchangedLong-8                    96.0          39.7          -58.65%
BenchmarkUnchangedSimple-8                  41.2          15.2          -63.11%
BenchmarkModifiedUnicode-8                  36.6          59.0          +61.20%
BenchmarkModifiedLong-8                     86.7          118           +36.10%
BenchmarkModifiedLongSpecialChars-8         135           161           +19.26%
BenchmarkModifiedSimple-8                   43.7          62.1          +42.11%
BenchmarkModifiedUnicode2-8                 73.0          110           +50.68%
BenchmarkLeadingUnderscoresDigitUpper-8     39.2          60.9          +55.36%
BenchmarkDigitUpper-8                       34.4          42.5          +23.55%
BenchmarkDigitUpper2-8                      85.2          114           +33.80%
```

so `b` when using `b := make([]byte, 0, 64)` doesn't actually escape to the heap because it never leaves the function nor does it cross any package boundaries:
`go build -gcflags '-m'`
```
./snake.go:85:6: can inline isLower
./snake.go:89:6: can inline isDigit
./snake.go:81:6: can inline isUpper
./snake.go:77:6: can inline isAlphanumeric
./snake.go:78:16: inlining call to isLower
./snake.go:78:30: inlining call to isUpper
./snake.go:78:44: inlining call to isDigit
./snake.go:18:13: inlining call to isLower
./snake.go:24:20: inlining call to isDigit
./snake.go:26:78: inlining call to isLower
./snake.go:26:99: inlining call to isDigit
./snake.go:42:12: inlining call to isUpper
./snake.go:43:31: inlining call to isUpper
./snake.go:43:50: inlining call to isDigit
./snake.go:48:31: inlining call to isLower
./snake.go:48:50: inlining call to isDigit
./snake.go:55:21: inlining call to isAlphanumeric
./snake.go:55:21: inlining call to isLower
./snake.go:55:21: inlining call to isUpper
./snake.go:55:21: inlining call to isDigit
./snake.go:64:31: inlining call to isUpper
./snake.go:64:50: inlining call to isDigit
./snake.go:69:31: inlining call to isLower
./snake.go:69:50: inlining call to isDigit
./snake.go:7:16: leaking param: s to result ~r1 level=0
./snake.go:74:15: string(b) escapes to heap
./snake.go:39:11: Snakecase make([]byte, 0, 64) does not escape
./snake.go:44:37: index bounds check elided
./snake.go:65:37: index bounds check elided
```

but using the unsafe.Pointer causes it to be heap allocated and adversely affects performance:
```
./snake.go:88:6: can inline isLower
./snake.go:92:6: can inline isDigit
./snake.go:84:6: can inline isUpper
./snake.go:80:6: can inline isAlphanumeric
./snake.go:81:16: inlining call to isLower
./snake.go:81:30: inlining call to isUpper
./snake.go:81:44: inlining call to isDigit
./snake.go:20:13: inlining call to isLower
./snake.go:26:20: inlining call to isDigit
./snake.go:28:78: inlining call to isLower
./snake.go:28:99: inlining call to isDigit
./snake.go:44:12: inlining call to isUpper
./snake.go:45:31: inlining call to isUpper
./snake.go:45:50: inlining call to isDigit
./snake.go:50:31: inlining call to isLower
./snake.go:50:50: inlining call to isDigit
./snake.go:57:21: inlining call to isAlphanumeric
./snake.go:57:21: inlining call to isLower
./snake.go:57:21: inlining call to isUpper
./snake.go:57:21: inlining call to isDigit
./snake.go:66:31: inlining call to isUpper
./snake.go:66:50: inlining call to isDigit
./snake.go:71:31: inlining call to isLower
./snake.go:71:50: inlining call to isDigit
./snake.go:9:16: leaking param: s to result ~r1 level=0
./snake.go:41:11: make([]byte, 0, 64) escapes to heap
./snake.go:76:35: Snakecase &b does not escape
./snake.go:46:37: index bounds check elided
./snake.go:67:37: index bounds check elided
```
```
benchmark                                   old ns/op     new ns/op     delta
BenchmarkUnchangedLong-8                    96.0          39.4          -58.96%
BenchmarkUnchangedSimple-8                  41.2          14.7          -64.32%
BenchmarkModifiedUnicode-8                  36.6          49.0          +33.88%
BenchmarkModifiedLong-8                     86.7          88.4          +1.96%
BenchmarkModifiedLongSpecialChars-8         135           133           -1.48%
BenchmarkModifiedSimple-8                   43.7          48.9          +11.90%
BenchmarkModifiedUnicode2-8                 73.0          86.6          +18.63%
BenchmarkLeadingUnderscoresDigitUpper-8     39.2          47.5          +21.17%
BenchmarkDigitUpper-8                       34.4          40.4          +17.44%
BenchmarkDigitUpper2-8                      85.2          88.9          +4.34%

benchmark                                   old MB/s     new MB/s     speedup
BenchmarkUnchangedLong-8                    343.66       838.21       2.44x
BenchmarkUnchangedSimple-8                  266.95       748.43       2.80x
BenchmarkModifiedUnicode-8                  245.89       183.64       0.75x
BenchmarkModifiedLong-8                     334.32       328.22       0.98x
BenchmarkModifiedLongSpecialChars-8         222.20       224.13       1.01x
BenchmarkModifiedSimple-8                   251.79       225.08       0.89x
BenchmarkModifiedUnicode2-8                 355.94       300.10       0.84x
BenchmarkLeadingUnderscoresDigitUpper-8     153.03       126.21       0.82x
BenchmarkDigitUpper-8                       145.42       123.74       0.85x
BenchmarkDigitUpper2-8                      199.58       191.24       0.96x

benchmark                                   old allocs     new allocs     delta
BenchmarkUnchangedLong-8                    1              0              -100.00%
BenchmarkUnchangedSimple-8                  1              0              -100.00%
BenchmarkModifiedUnicode-8                  1              1              +0.00%
BenchmarkModifiedLong-8                     1              1              +0.00%
BenchmarkModifiedLongSpecialChars-8         1              1              +0.00%
BenchmarkModifiedSimple-8                   1              1              +0.00%
BenchmarkModifiedUnicode2-8                 1              1              +0.00%
BenchmarkLeadingUnderscoresDigitUpper-8     1              1              +0.00%
BenchmarkDigitUpper-8                       1              1              +0.00%
BenchmarkDigitUpper2-8                      1              1              +0.00%

benchmark                                   old bytes     new bytes     delta
BenchmarkUnchangedLong-8                    48            0             -100.00%
BenchmarkUnchangedSimple-8                  16            0             -100.00%
BenchmarkModifiedUnicode-8                  3             64            +2033.33%
BenchmarkModifiedLong-8                     48            64            +33.33%
BenchmarkModifiedLongSpecialChars-8         32            64            +100.00%
BenchmarkModifiedSimple-8                   16            64            +300.00%
BenchmarkModifiedUnicode2-8                 3             64            +2033.33%
BenchmarkLeadingUnderscoresDigitUpper-8     5             64            +1180.00%
BenchmarkDigitUpper-8                       5             64            +1180.00%
BenchmarkDigitUpper2-8                      32            64            +100.00%
```

it looks like the way it is is the most efficient.

Comment (achille-roussel) [reply snake.go:39]: Nice! 👍

File: snake.go
Comment (achille-roussel) [inline snake.go]:
```suggestion
var asciiLowercaseArray = [256]byte{
```
This will add a second check to ensure that we cover all possible cases, if we were off-by-one the compiler would be inserting bound checks where we access the array.

Comment (deankarn) [reply snake.go]: love it! thanks!

**Review by achille-roussel — APPROVED** [review]
Seems like you've done very strong testing and we are going to see significant upsides with this change. The algorithm does become way more complex than it used to be, so I'd love for someone else to also review and approve the PR to get another point of view on the subject.

Next step: assembly? https://github.com/mmcloughlin/avo

**Comment by deankarn** [comment]
@achille-roussel thanks 👍 

agree with having more than one approver and opinion!

Next step: https://www.rust-lang.org/ ;)
(although my rust version does not seem to be quite as fast, Go does a great job with string manipulation)

**Review by dominicbarnes — APPROVED** [review]

**Comment by deankarn** [comment]
@achille-roussel @dominicbarnes do we want anyone else to take a look before merging?

**Comment by achille-roussel** [comment]
I'm good 👍

**Comment by achille-roussel** [comment]
Nice work on this by the way 💯

**Pull request merged**
